### PR TITLE
Proof of Concept

### PR DESCRIPTION
**Pull request by newsch**
This is an initial implementation of searching for Wikipedia urls/Wikidata QIDs and simplifying html.
It's a little messy, inefficient, ~~and doesn't handle multiple languages or writing to an output directory~~.

Remaining work for this PR:
- [x] Refactor into library/modules
- [x] Set up comparison of html processing with python implementation
  - processing of same html files
  - scraped vs enterprise dump
- [x] Add unit tests
- [x] Revisit `WikipediaTitleNorm` design (parsing 3 million QIDs is almost instant, 1.5 million titles around a minute)
- [x] Add logging/tracing
- [x] Add proper CLI
- [x] Look through TODOs
- [x] Add multiple language support
- [x] Write to directory

### CODE REVIEW

**Review by biodranik — COMMENTED** [review]
1. Does it make sense to start splitting the code into logical modules in separate sources?
2. Implementing convenient cmd line args may save some time.
3. Some common data preparation tasks or common test scenarios can be also implemented in rust with a cmd line param if it saves time in the longer term.
4. Is there a Rust library to read and extract tags from planet.o5m files directly? Just curious, using osmfilter or osmium-tool is also ok.

**Comment by newsch** [comment]
> 1. Does it make sense to start splitting the code into logical modules in separate sources?

I think so, at the very least it needs a `lib.rs` to make a separate binary for testing the html simplifying. I won't go overboard :wink:.

> 2. Implementing convenient cmd line args may save some time.

I agree, but I wanted to get the logic working before worrying about that.

> 3. Some common data preparation tasks or common test scenarios can be also implemented in rust with a cmd line param if it saves time in the longer term.

Good point, I'll keep that in mind if I run into any other work that needs to be done.

> 4. Is there a Rust library to read and extract tags from planet.o5m files directly? Just curious, using osmfilter or osmium-tool is also ok.

There are [a number of crates related to the `.pbf` format](https://lib.rs/search?q=osm+OR+openstreetmap), and bindings for libosmium, but I haven't come across any `.o5m` ones that seem thoroughly developed.

**Comment by biodranik** [comment]
Can pbf be used to get fresh OpenStreetMap.org updates without converting it to o5m?

**Comment by newsch** [comment]
The planet downloads are available in pbf: https://planet.openstreetmap.org/pbf/

I think this is saying you can apply the daily/hourly diffs to pbf, but I'm not sure I'm reading it correctly: https://wiki.openstreetmap.org/wiki/Planet.osm/diffs#Using_the_replication_diffs

This makes it seem like diffs can be appended at the end of the file, which seems like you'd need to read the entire file to get the true state of any node/object?
https://wiki.openstreetmap.org/wiki/PBF_Format#What_are_the_replication_fields_for?

I don't understand enough about these formats or how the generator works to know if they'd be compatible.

**Comment by biodranik** [comment]
That's why we use o5m. It is suitable for incremental updates, and osmupdate tool creates a new planet.o5m file automatically by applying these updates.

**Comment by newsch** [comment]
It looks like osmupdate supports pbf for the same thing, but I don't understand how that works

**Comment by biodranik** [comment]
Tested, it works for pbf too. Cool! The whole planet can be stripped to 58gb:
`"$OSMUPDATE" -v --drop-authors --drop-version --hash-memory=16000 planet-230529.osm.pbf planet-updated.osm.pbf`

**Comment by newsch** [comment]
Alright, here's what I've learned from comparing the html outputs:

The simplifying step of the python and the rust versions give comparable output (proper minification is still needed), but the html that the python gets from the api is much more simple than I expected.

It uses [the Extracts API](https://www.mediawiki.org/wiki/Extension:TextExtracts), which strips most of the markup.

The html in the dumps on the other hand seem much closer to the content in a complete article. Size-wise the dump html is around 10x the size of that of the extracts for the subset I looked at.

To get to parity with that output, we'll need to add additional steps to the html processing to remove:
- mw-specific tags and attributes
- links
- images
- [info boxes](https://en.wikipedia.org/wiki/Help:Infobox)

This is doable, it will involve some more exploration of what exactly to match on.

I'll make issues for that and the minification.

**Comment by biodranik** [comment]
Minification is not a blocker and can be done later, if necessary. It would be great to compare final mwm sizes, AFAIR, some compression is used there. 

Showing images for those who wants them, and maybe leave links too may be a good idea.

**Comment by newsch** [comment]
> Minification is not a blocker and can be done later, if necessary.

:thumbsup:

> It would be great to compare final mwm sizes, AFAIR, some compression is used there.

Good point. I've seen the compression [referenced in `text_storage.hpp`](https://github.com/organicmaps/organicmaps/blob/acc7c0547db4285dd8841ae7f98811268e38d908/coding/text_storage.hpp#L28).
I'll look into doing this locally.

> Showing images for those who wants them, and maybe leave links too may be a good idea.

I know that's a goal, but right now the links/images are all relative urls so they'll still need some processing.

**Comment by newsch** [comment]
Next steps are:
- reviewing html output (#4)
- connecting output to map generator (mainly handling wikipedia -> wikidata mapping)
- initial profiling of pipeline (right now it can roughly keep up with tar on my machine, and that's with writing the output files to a ramdisk)

After that I think this will be ready to try out in production!

File: src/html.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            "References",
+        },
+       "es" => set!{
+            "Vínculos de interés",
+            "Véase también",
+            "Enlaces externos",
+            "Referencias"
+        },
+        "fr" => set!{
+            "Bibliographie",
```
Comment (biodranik) [inline src/html.rs:34]: Is there a more robust way to exclude some sections for all languages?

Comment (biodranik) [reply src/html.rs:34]: No need to collapse, a config looks like a better option (keep adding many other languages in mind, and potential contributors who doesn't read rust code)

Comment (newsch) [reply src/html.rs]: I added a compile-time json config in 7e6b39a, adding in a flag for loading one at runtime is straightforward if we want to do that later. Using a different config language is also a quick change.

File: src/html.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        if element.text().all(|t| t.trim().is_empty()) {
+            to_remove.push(element.id());
+        }
+    }
+
+    for id in to_remove.drain(..) {
+        if let Some(mut node) = document.tree.get_mut(id) {
+            node.detach();
+        }
+    }
```
Comment (biodranik) [inline src/html.rs:79]: Can copy-paste be avoided?

File: src/html.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    document.html()
+}
+
+#[cfg(test)]
+mod test {
```
Comment (biodranik) [inline src/html.rs:85]: Is it hard to make a simple test for the function above?

File: src/wm/mod.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    };
+
+    ids.get(&wikidata_id).map(|_| wikidata_id)
+}
+
+pub fn is_wikipedia_match(
+    titles: &HashSet<WikipediaTitleNorm>,
+    page: &Page,
+) -> Option<WikipediaTitleNorm> {
+    // TODO: handle multiple languages
```
Comment (biodranik) [inline src/wm/mod.rs:61]: 1. Is it in English only now?
2. Does it make sense to test this function?

File: src/wm/mod.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+/// - titles `Spatial Database`
+/// - urls `https://en.wikipedia.org/wiki/Spatial_database#Geodatabase`
+/// - osm-style tags `en:Spatial Database`
+///
+/// ```
+/// use om_wikiparser::wm::WikipediaTitleNorm;
+///
+/// let title = WikipediaTitleNorm::from_title("Article Title", "en");
+/// let url = WikipediaTitleNorm::from_url("https://en.wikipedia.org/wiki/Article_Title#Section").unwrap();
+/// assert_eq!(url, title);
```
Comment (biodranik) [inline src/wm/mod.rs:113]: Does it make sense to test it?

File: src/wm/mod.rs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+///
+/// See https://www.wikidata.org/wiki/Wikidata:Glossary#QID
+///
+/// ```
+/// use std::str::FromStr;
+/// use om_wikiparser::wm::WikidataQid;
+///
+/// let with_q = WikidataQid::from_str("Q12345").unwrap();
+/// let without_q = WikidataQid::from_str("12345").unwrap();
+/// assert_eq!(with_q, without_q);
```
Comment (biodranik) [inline src/wm/mod.rs:89]: Does it make sense to test it?

**Review by biodranik — APPROVED** [review]
LGTM with a few comments

**Comment by newsch** [comment]
Thanks, I'll address those comments, rebase into individual changes, and merge.

File: src/wm/mod.rs (hunk 4 above)
Comment (newsch) [reply src/wm/mod.rs:61]: Oops, that comment slipped through when I handled multiple languages in 6e5385d. I'll remove it.

> 1. Is it in English only now?

Currently it will work with any language, but it only processes a single dump at a time. So when it reads an english dump, each article json has a `.in_language.identifier` field set to `en`, and the program writes that html to a `QXXXXX/en.html` file.

Running the program multiple times with different language dumps will fill in the various `QXXXXX/$lang.html` files.
We could extend it to process multiple dumps in parallel, but I don't expect there to be much of a speedup right now.

> 2. Does it make sense to test this function?

The doctest on the type definition verifies that the two constructors parse and normalize correctly, but doesn't check for various error cases.
Do you think there should be more?

Comment (biodranik) [reply src/wm/mod.rs:61]: 1. Does it imply a wrapping script that launches the app for each language/file? That is ok (and can be paralleled in bash by launching several processes simultaneously), but should be documented. Or is there a better approach?
2. It may make sense to check values that are in OSM. Users can make a lot of mistakes.

Comment (newsch) [reply src/wm/mod.rs:61]: > 1. Does it imply a wrapping script that launches the app for each language/file? That is ok (and can be paralleled in bash by launching several processes simultaneously), but should be documented. Or is there a better approach?

I haven't tried it out yet, but there are a couple of options that come to mind:
- Decompress the archives serially so they're all concatenated together into stdin. This looks possible with `gunzip`/`tar`, not sure about python `pgzip`.
- As you say, run the program repeatedly through a wrapper script, using a for loop, `xargs`, `parallel`, etc.
- Pass the decompression command to the program and have it spawn the subprocess directly, it could do this in parallel and pass the results to the same worker pool.

> 2. It may make sense to check values that are in OSM. Users can make a lot of mistakes.

Understood. I have been using the world list of urls/ids from the map generator with no problems, but if we switch to using OSM data directly I'll rethink this. The program will log any issues it has parsing titles/QIDs.

File: src/html.rs (hunk 3 above)
Comment (newsch) [reply src/html.rs:85]: No, I just didn't bother since I'll be changing this in the next PR. I can add some if you'd like.

File: src/html.rs (hunk 2 above)
Comment (newsch) [reply src/html.rs:79]: I'll be improving and refactoring this in the next PR, if this bit survives I'll move it to a function.

File: src/wm/mod.rs (hunk 5 above)
Comment (newsch) [reply src/wm/mod.rs:131]: I added some checks for whitespace, empty strings, and tests for errors in 70f7edf, is there something else you think should be handled?

File: src/html.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+pub fn simplify(html: &str, lang: &str) -> String {
+    let mut document = Html::parse_document(html);
+
+    let mut to_remove = Vec::new();
+
+    // remove sections
+
+    if let Some(bad_sections) = CONFIG.bad_sections.get(lang) {
+        for header in document.select(&HEADERS) {
+            // TODO: should this join all text nodes?
```
[the remark this answers is not in the document]
Comment (newsch) [reply src/html.rs]: I need to look through a good sample of the articles and check for things like formatting in headers, tags that are used, and which wikimedia meta tags are responsible for a third of the document size.
That's part of the next work.

File: src/wm/mod.rs (hunk 6 above)
Comment (newsch) [reply src/wm/mod.rs:100]: Same as [above](https://github.com/organicmaps/wikiparser/pull/3#discussion_r1239061793).

File: src/wm/mod.rs (hunk 4 above)
Comment (biodranik) [reply src/wm/mod.rs]: It should be ok to run tasks in parallel using bash for loop. We'll rethink it if necessary later.

**Review by biodranik — APPROVED** [review]
